[lldb/Interpreter] Add requirements to Scripted Interface abstract methods

medismailben opened the pull request:
This patch adds new requirements to the Scripted Interface abstract method checker to check the minimum number of argument for abstract methods.

This check is done when creating the interface object so the object is not created if the user implementation doesn't match the abstract method requirement.

bulbazord commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:
```diff
@@ -78,8 +100,11 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
-    auto create_error = [](std::string message) {
-      return llvm::createStringError(llvm::inconvertibleErrorCode(), message);
+    Log *log = GetLog(LLDBLog::Script);
+    auto create_error = [](llvm::StringRef format, auto &&...ts) {
```
I think it would be better to be specific with the type here: `const char *`.
`llvm::StringRef` may or may not point to a null-terminated string so calling `.data()` on the parameter sets off alarm bells in my head. I can manually verify that this is never called with anything other than a `const char *` but somebody else may not be so careful in the future.

medismailben replied:
Made it an `llvm::StringLitteral` instead ;)

bulbazord commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:
```diff
@@ -186,36 +210,49 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
+        abstract_method_errors = llvm::joinErrors(
+            std::move(abstract_method_errors),
+            std::move(create_error(
+                "Abstract method {0}.{1} has unexpected argument count.",
```
I don't think this error is actionable as-is. Is there a way you can include the argument count and the minimum number of arguments?

medismailben replied:
Yeah, it would just require more plumbing.

medismailben committed:
[lldb/Interpreter] Add requirements to Scripted Interface abstract methods

This patch adds new requirements to the Scripted Interface abstract
method checker to check the minimum number of argument for abstract
methods.

This check is done when creating the interface object so the object is
not created if the user implementation doesn't match the abstract method
requirement.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>
